### DISCUSSION TITLE
How to combine layers in onnx

### DISCUSSION THREAD

**Discussion by washraf** [onnx/onnx#2953] [Q&A]
# Ask a Question
I want to merge layers of two models that share the first k layers.
### Question
I want to merge/combine two models as shown in the image below. This merging has a lot of benefits in terms of computation time.

![fuse drawio](https://user-images.githubusercontent.com/4272790/89742736-933f2900-da6a-11ea-8035-a099e48baa4f.png)

This technique is available in almost all deep learning frameworks such as in [Pytorch](https://discuss.pytorch.org/t/combining-trained-models-in-pytorch/28383) and [Keras](https://datascience.stackexchange.com/questions/26103/merging-two-different-models-in-keras). 

### Further information
- Relevant Area (*e.g. model usage, backend, best practices, converters, shape_inference, version_converter, training, test, operators*): 

- Is this issue related to a specific model?  
**Model name** (*e.g. mnist*):  
**Model opset** (*e.g. 7*):

### Notes
Any additional information, code snippets.

**Comment by liqunfu** [comment]
Thanks @washraf, can you provide a PyTorch example with shared modules that cannot be exported into ONNX as expected?

**Comment by washraf** [comment]
Hello @liqunfu, I don't have an issue in the conversion of PyTorch models. My goal is to combine two onnx models that share most of their layers into one model. The reference of both Keras and PyTorch is to show that the concept is doable in other formats. 

For example, assume I have two instances of ResNet and they share the same layers from 1 to k, and the rest of the layers each onnx model have its own layers. I want to combine the first layers, to gain many advantages in computation time.

**Comment by hariharans29** [comment]
Please see if you can start with using this: https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md#creating-an-onnx-model-using-helper-functions. 

Essentially you need to create a "new" model - but you should be able to load both the models and try and connect the layers by using the exposed ONNX APIs.
